sound/soc/sof/ipc4-pcm.c

```diff
@@ -332,12 +332,12 @@ static int sof_ipc4_trigger_pipelines(struct snd_soc_component *component,
 	 * guaranteed for each fork independently.
 	 */
 	if (state == SOF_IPC4_PIPE_RUNNING || state == SOF_IPC4_PIPE_RESET)
-		for (i = pipeline_list->count - 1; i >= 0; i--) {
+		for (i = 0; i < pipeline_list->count; i++) {
 			spipe = pipeline_list->pipelines[i];
 			sof_ipc4_add_pipeline_to_trigger_list(sdev, state, spipe, trigger_list);
 		}
 	else
-		for (i = 0; i < pipeline_list->count; i++) {
+		for (i = pipeline_list->count - 1; i >= 0; i--) {
```

Comment:
I am not following why we did all this and why the 'new order' is better.

Surely there was something that led us to go from source to sink initially?

Reply:
@plbossart I think the key point I missed in the explanation is that the order of widget setup doesnt matter. But what really matters is the order of pipeline trigger and this is determined by the order in which widgets are set up. So when the widgets are set up in the order source->sink, pipelines are added to the trigger list in the same order and we trigger in the reverse order to honor the sink->source requirement with IPC4.

So what happens today is that if you had 2 DAI's for capture (ie DMIC capture and SSP capture for reference as in the example in my commit message), the source to sink widget setup order adds the pipelines in the order 1->2->0. 
So triggering in reverse would mean the reference capture pipeline goes first. This fails with an IPC timeout in the firmware. What we really want is for the pipeline 2 to be triggered first.

Reply:
still struggling to figure out why we didn't use the same order for setup and trigger. Is this because it was less complicated to do so initially to follow the source->sink links?

also isn't the IPC timeout in the firmware an issue? Shouldn't there be some sort of state check to discard IPCs that make no sense?

Reply:
We could have (and should have) used the sink->source for set up earlier but the reason we didnt is likely that we never had a situation where we have multiple sources/sinks in the list of DAPM widgets before and so it was convenient to just to set up in the source->sink order and just pink the reverse order for triggering.

The IPC timeout is absolutely a firmware issue and I am going to debug that as well but I think this PR makes itthe triggering order more logical anyway

Reply:
I see what you mean now, for the capture part we indeed had a single source split in different streams. The AEC reference coming from a different DAI is a new configuration. Maybe that's something we should add to the nocodec topology by tying one SSP dai pipeline to an AEC pretend module.

Reply:
@ranj063 This clarification addressed my comment. It's also worth checking whether https://github.com/thesofproject/sof/pull/8084 (just merged) has impact to FW behaviour.

Reply:
@kv2019i I will check with 8084 once I add the mock aec to nocodec. But this PR is still in the right direction I think.

Reply:
@kv2019i : this alone did not help.
This PR is very much needed and its critical for the AEC feature to work on the chromebook